New python manual

stefano-ottolenghi opened the pull request:
Apart from a couple of sentences and examples, the manual has been entirely rewritten.

A short summary of intentions:
- create a journey, so that people reading the manual from the first to the last page will see progressively more involved examples.
- make pages and sections as much self-contained as possible, so that people can land on whatever section and still get what they want out of it.
- make the experience practical, by showing as much examples and code as possible, and avoiding telling.
- provide a gentle entrypoint into Cypher → tried to avoid using always the same example, and instead used different clauses and operators.
- cover points that are often source of confusion or error, or points that have been requested by users
- make the manual as useful/friendly for the 80% of use cases, even at the cost of removing some information useful to the remaining 20%, and refer to API docs for advanced reference.

stefano-ottolenghi committed:
Final draft of new python manual

neo-technology-commit-status-publisher commented:
Looks like you've updated the documentation!

Check out your changes at https://neo4j-docs-drivers-372.surge.sh
<!-- Sticky Pull Request Comment -->

stefano-ottolenghi committed:
polish

stefano-ottolenghi committed:
unify install and get an instance sections in get started

stefano-ottolenghi committed:
change section title

stefano-ottolenghi committed:
replace legacy syntax for tags

nvitucci commented on python-manual/modules/ROOT/pages/async.adoc:
```diff
@@ -0,0 +1,41 @@
+----
+
+[TIP]
+Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although asyncio is often easier to implement in an application.
```
```suggestion
Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although `asyncio` is often easier to implement in an application.
```

nvitucci commented on python-manual/modules/ROOT/pages/async.adoc:
```diff
@@ -0,0 +1,41 @@
+= Async query flow
+
+To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.
```
```suggestion
To interact with the database in an asynchronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the synchronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.
```

robsdedude replied:
```suggestion
To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.
```

`AsyncGraphDatabase` is not callable.

nvitucci commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
+session.run("MATCH (p:Person {name: " + name + "}) RETURN p")
+----
+
+- *Specify node labels* in all queries. To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cyper -- Label expressions].
```
```suggestion
- *Specify node labels* in all queries. To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cypher -- Label expressions].
```

nvitucci commented on python-manual/modules/ROOT/pages/data-types.adoc:
```diff
@@ -0,0 +1,411 @@
+= Data types and mapping to Cypher types
+
+The tables in this section show the mapping between Cyper data types and Python types.
```
```suggestion
The tables in this section show the mapping between Cypher data types and Python types.
```

nvitucci reviewed (commented):
Some typos

robsdedude commented on python-manual/modules/ROOT/pages/async.adoc:
```diff
@@ -0,0 +1,41 @@
+= Async query flow
+
+To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.
```
```suggestion
To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define all functions that should be awaited  as `async`.
```

stefano-ottolenghi replied:
I think it's clearer the way it is, so that `async` is as close to `define` as possible.

robsdedude commented on python-manual/modules/ROOT/pages/async.adoc:
```diff
@@ -0,0 +1,41 @@
+[source,python, role=nocollapse]
+----
+import asyncio
+from neo4j import AsyncGraphDatabase
```
```suggestion

from neo4j import AsyncGraphDatabase
```

:grimacing: the nitpicking... I can't help it, sorry.

robsdedude commented on python-manual/modules/ROOT/pages/async.adoc:
```diff
@@ -0,0 +1,41 @@
+    URI = "<URI for Neo4j database>"
+    AUTH = ("<Username>", "<Password>")
+
+    async with AsyncGraphDatabase.driver(URI, auth=AUTH) as driver:
```
```suggestion
    uri = "<URI for Neo4j database>"
    auth = ("<Username>", "<Password>")

    async with AsyncGraphDatabase.driver(uri, auth=auth) as driver:
```
I'd only all-caps global constants.

stefano-ottolenghi replied:
Moved out of main for consistency with other examples

robsdedude commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,99 @@
+= Coordinate different transactions with bookmarks
+
+When working with a Neo4j cluster, <<causal_consistency>> is enforced by default in most cases, which guarantees that a query is able to read changes made by previous queries. However, if you xref:transactions.adoc[take control of transactions] or if your application is running concurrent transactions, you might need to use _bookmarks_ to avoid the cluster falling back to <<eventual_consistency>> ...
```
I'm not sure this conveys the right message. Causal consistency is enforced when running queries that required causal consistency in the same session. That's all the driver currently offers (until the bookmark manager is a thing). On the other side we communicate that sessions are cheap to create and that users can and probably should wrap each query in it's own session. That is, unless they explicitly require causal consistency.

stefano-ottolenghi replied:
Good point. Shall we just take away the note on sessions being cheap to create? Let's just encourage them to use the same session for as much work as possible, I don't think there's any harm in that, right?

robsdedude replied:
There is harm in that when they scale their application and go cluster (which admittedly is pretty far down the experience route), they'll have all their queries in one or only a few causal chains. Which makes their application stall a lot while the followers wait to catch up with the bookmarks. Ideally you should have every query in its own session unless you need causal chaining.

stefano-ottolenghi replied:
I've made another attempt, hopefully the last.

robsdedude commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,115 @@
+
+== Custom address resolver
+
+You can specify a custom _resolver_ function to resolve host and port values ahead of DNS resolution. This function is called with a 2-tuple of `(host, port)` and should return an iterable of 2-tuples `(host, port)`.
```
Not 2-tuples, but `neo4j.IPv4Address` or `neo4j.IPv6Address` objects which indeed inherit from tuple but a) have more methods and b) have 2 elements from IPv4 and 4 for IPv6. 

It then can return an iterable of (any combination of) `tuple`, `neo4j.Address` and its subtypes. In fact, anything you could pass into `neo4j.Address`.

And with that, I just realized that this is a copy paste of the API docs *insert sinking titanic picture*.

robsdedude replied:
Looking at the driver's code, the `yield "::1", 7687` in the example might even not be doing what it suggests to do. I'll have a look at this. For now, you can leave it as is and I'll open a PR to the manual or give you a ping when I update the API docs.

stefano-ottolenghi replied:
Ahah yeah exactly, this is a hard copy-paste that I didn't even test I think, tbh. I'll leave this on and look at it when I can

stefano-ottolenghi replied:
Addressed in #374.

robsdedude commented on python-manual/modules/ROOT/pages/data-types.adoc:
```diff
@@ -0,0 +1,411 @@
+from neo4j import GraphDatabase
+from datetime import datetime
+import pytz
+from neo4j.time import DateTime
```
```suggestion
from datetime import datetime

from neo4j import GraphDatabase
from neo4j.time import DateTime
import pytz
```

robsdedude commented on python-manual/modules/ROOT/pages/data-types.adoc:
```diff
@@ -0,0 +1,411 @@
+
+# Python's native datetimes work as well.
+# They don't support the full feature-set of Neo4j's type though.
+# py_friends_since = datetime(year=1999, month=11, day=23, hour=7, minute=47)
```
```suggestion
# py_friends_since = datetime(year=1999, month=11, day=23, hour=7, minute=47,
#                             microsecond=4) 
```
maybe to better match the above `DateTime`?

stefano-ottolenghi replied:
Since it wouldn't match exactly anyway (because of the 123 nanoseconds difference), I think we can skip the extra parameter. Reopen if you don't feel so.

robsdedude commented on python-manual/modules/ROOT/pages/data-types.adoc:
```diff
@@ -0,0 +1,411 @@
+from neo4j.time import Time
+
+t = Time(hour=7, minute=47, nanosecond=4123, tzinfo=pytz.FixedOffset(-240))
+print(t)  # '07:47:00.000004000-04:00'
```
```suggestion
print(t)  # '07:47:00.000004123-04:00'
```
I assume. And other instances of the same below.

stefano-ottolenghi replied:
Ops. I think it came from the output of the native type conversion, and I mixed them up. What an eye!

robsdedude commented on python-manual/modules/ROOT/pages/data-types.adoc:
```diff
@@ -0,0 +1,411 @@
+
+[source, python]
+----
+from neo4j.time import DateTime
```
```suggestion
from neo4j.time import DateTime
import pytz
```

robsdedude commented on python-manual/modules/ROOT/pages/data-types.adoc:
```diff
@@ -0,0 +1,411 @@
+[source, python, indent=0]
+----
+from neo4j.time import Time
+
```
I assume you purposefully only used one blank line for the really short examples. Looks reasonable to me.

robsdedude commented on python-manual/modules/ROOT/pages/index.adoc:
```diff
@@ -1,51 +1,83 @@
+from neo4j import GraphDatabase
 
-link:https://graphacademy.neo4j.com/courses/app-python/?ref=guides[Learn more^,role=button]
+URI = "<URL for Neo4j database>"
```
```suggestion
URI = "<URI for Neo4j database>"
```

robsdedude commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
+driver.session()
+----
+
+// - Execute as many queries within the same session. Minimise the number of transactions?
```
It's less about reducing the number of sessions but reducing the number of transactions. Using batching reduces the number or round-trip times, reduces the impact of the overhead on the server to parse the cypher query, start a transaction, potentially locking, and more.

robsdedude replied:
I just saw you mention this further down. So maybe this comment can be :wastebasket:ed.

stefano-ottolenghi replied:
Yeah, they are related. Although I see it as slightly different, in that the note after is about batching creation, while we also want to encourage general parsimony in transactions. They boil down to the same principle, but it might not be so apparent to new users, so I wanted to be extra clear that it is meaningful also in general, not just when inserting lots of data.

robsdedude replied:
Fair points. It's also just a comment. Let me highlight again, though, that using fewer sessions doesn't give you much (other than more automatic causal chaining which might even end up slightly slowing down the application if it's not needed). You can resolve this comment, when you've read this.

robsdedude commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
+
+// - Execute as many queries within the same session. Minimise the number of transactions?
+
+- *Use xref:query-simple#query-parameters[query parameters]* instead of hardcoding or concatenating values into queries.
```
Is it worth explaining why this is good practice? I.e., not only for security, but also so that the DMBS can cache how to parse and execute the query.

stefano-ottolenghi replied:
It is mentioned in the `query parameters` section that's linked there. I can add a "This allows the database to cache queries" or something similar though

robsdedude commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
+numbers = [{"value": random()} for _ in range(10000)]
+session.run("""
+    WITH $numbers as batch
+    UNWIND batch as node
```
```suggestion
    WITH $numbers AS batch
    UNWIND batch AS node
```

robsdedude commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
+    session.run("MERGE (:Number {value: $value})", value=random())
+----
+
+- *Split nodes and relationships creation when inserting a lot of records*.
```
I didn't know that this is more efficient? Why is that? Is it worth mentioning this here?

stefano-ottolenghi replied:
I can't come up with a good reason why it should be more efficient. I remember reading/hearing it somewhere (that I can't find again now) and then testing it to verify it. Now I discovered that it only seems to give a performance boost on a cold database, but if I run the test twice times becomes comparable. I'll investigate a bit.

robsdedude commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
+[source,python,role=noheader]
+----
+# Good practice
+driver.execute_query("MATCH (p:Person) RETURN p", routing_="r")
```
```suggestion
```
too soon :sob: 

(also above and below)

stefano-ottolenghi replied:
It was commented out, but now taken away entirely

robsdedude replied:
Ah sorry. My ascii docs foo isn't particularly strong.

robsdedude commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,201 @@
++-----------------+----------------+----------------+---------------------+
+
+Total database accesses: ?
+----
```
```suggestion
Total database accesses: ?
"""
----
```

robsdedude commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,119 @@
+    session.run("CREATE (a:Person {name: $name})", name=name)
+----
+
+An auto-commit transaction will be committed _at the latest_ when the session is destroyed, or before another transaction is executed within the same session. Other than that, there is no clear guarantee on when exactly an auto-commit transaction will be committed during the lifetime of a session. Auto-commit transactions should only be used when xref:transactions#managed-transactions[transaction ...
```
You might want to mention that `result.consume()` also gives you the guarantee that the tx has been committed.

robsdedude commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,119 @@
+with driver.session() as session:
+    query = Query("CREATE (a:Person {name: $name})",
+                  timeout=1.0,
+                  metadata={app_name: "people"})
```
```suggestion
                  metadata={"app_name": "people"})
```

robsdedude commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,119 @@
+            "properties": {
+                property_key: "Alice",
+            },
+        }
```
Are you using a dict for the parameters here on purpose? Nothing against it, just to be clear, but I was surprised as it comes without a word about it.

stefano-ottolenghi replied:
Good point, I missed to update this after taking away the dict structure (or it came straight from Michael Simon's notes). Use kwargs now.

robsdedude commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,119 @@
+watch("neo4j", out=sys.stdout)
+
+
+----
```
```suggestion
----
import sys

from neo4j.debug import watch

watch("neo4j", out=sys.stdout)
----
```

robsdedude commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,233 @@
+[source,python]
+----
+def pandas_df(tx):
+    result = tx.run("UNWIND range(1, 10) AS n RETURN n, n+1 as m")
```
```suggestion
    result = tx.run("UNWIND range(1, 10) AS n RETURN n, n+1 AS m")
```

stefano-ottolenghi replied:
Same in API docs.

robsdedude replied:
Thanks for the hint.

robsdedude commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,289 @@
+= Fine-grained transaction control
```
This file seems to have some overlap with `python-manual/modules/ROOT/pages/query-simple.adoc` :thinking:

stefano-ottolenghi replied:
Indeed. This is my solution to the void generated by taking away `execute_query`. Without `execute_query`, there is no way to make queries without sessions and transactions and `Result` methods. Which is a lot to internalize before running even one query. So my solution is that, in the basic workflow, we pretend that managed transactions are just a code structure (almost boilerplate code) that you have to use to run queries, with not much justification. But at least it gets you started. Then we can revise the sloppy, black-box attitude in the subsequent pages. But I think that if I were to start explaining all the content of `transactions.adoc` before being able to do `MATCH (n:Person)`, the learning curve would just be too discouraging. So I have to skip over a lot of details. But I also want both pages to be somewhat self-contained, so that you can land on either and not be required to read the other. I'll try to do some more work to cross-reference them, but I hope this explains the rationale behind it at least :)

robsdedude replied:
That makes perfect sense. :ok_hand: People reading them in sequence will just need to skim over some parts they already know about. Better than leaving others puzzled, probably.

robsdedude commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,289 @@
+[#access-control-note]
+[NOTE]
+====
+Although executing a _write_ query under read mode will likely result in a runtime error, *you should not rely on `execute_read/write` for access control.* The difference between the two methods is that `read` transactions will be routed to any node of a cluster, whereas `write` ones will be directed to the _Leader_. However, you should not rely on this behavior for security purposes. In fact, de ...
```
The above section talks about `neo4j.READ_ACCESS`/neo4j.WRITE_ACCESS` and this note jumps to `execute_read/write`. While the same thing applies to both, this might not be evident to a newcomer reding this.

stefano-ottolenghi replied:
Agreed. Clarified.

robsdedude commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,289 @@
+
+[source,python]
+----
+with driver.session() as session:
```
```suggestion
with driver.session(database="neo4j") as session:
```
not necessary, but having the people copy paste snippets that implement best practices seems like a good idea to me.

robsdedude commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,289 @@
+[source,python]
+----
+import neo4j
+from os import sleep
```
```suggestion
from time import sleep

import neo4j
```

robsdedude commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,289 @@
+    except (DriverError, Neo4jError) as e:
+        if e.is_retryable() and attempts > 0:
+            sleep(1)
+            decrease_customer_balance(tx, customer_id, amount, attempts=attempts-1)
```
This will most likely not work because a failed `tx.run` will kill the transaction. You'd have to start a new transaction.

stefano-ottolenghi replied:
Ah, good point! I didn't test this because I needed to fill in some gaps, but I guess I'll have to and adjust it accordingly. Thanks

stefano-ottolenghi committed:
polish with feedback

stefano-ottolenghi committed:
method name

stefano-ottolenghi committed:
condense example

stefano-ottolenghi committed:
make more evident you need an instance
revise some performance notes, mention `neo4j-admin import database`
improve access control note
clarify sessions closing
move explicit transactions above session conf

stefano-ottolenghi committed:
fix examples

stefano-ottolenghi committed:
link pages a bit more together

stefano-ottolenghi committed:
improve sections openings
make menu titles more explicit
titles in bookmarks page
some rewording
more cross references

stefano-ottolenghi committed:
more clarification in bookmarks

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[glossary]
+[[LTS]]LTS:: A _Long Term Support_ release is one guaranteed to be supported for a number of years. Neo4j 4.4 is LTS, and Neo4j 5 will also have an LTS version.
```
```suggestion
[[LTS]]LTS:: A _Long Term Support_ release is one guaranteed to be supported for a number of years. 
Neo4j 4.4 is LTS, and Neo4j 5 will also have an LTS version.
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[glossary]
+[[LTS]]LTS:: A _Long Term Support_ release is one guaranteed to be supported for a number of years. Neo4j 4.4 is LTS, and Neo4j 5 will also have an LTS version.
+[[Aura]]Aura:: link:https://neo4j.com/cloud/platform/aura-graph-database/[Aura] is Neo4j's fully managed cloud service. It comes with both free and paid plans.
```
```suggestion
[[Aura]]Aura:: link:https://neo4j.com/cloud/platform/aura-graph-database/[Aura] is Neo4j's fully managed cloud service. 
It comes with both free and paid plans.
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[glossary]
+[[LTS]]LTS:: A _Long Term Support_ release is one guaranteed to be supported for a number of years. Neo4j 4.4 is LTS, and Neo4j 5 will also have an LTS version.
+[[Aura]]Aura:: link:https://neo4j.com/cloud/platform/aura-graph-database/[Aura] is Neo4j's fully managed cloud service. It comes with both free and paid plans.
+[[Driver]]Driver:: A link:https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Driver[`Driver`] object holds the details required to establish connections with a Neo4j database. Every Neo4j-backed application requires a `Driver` object.
```
```suggestion
[[Driver]]Driver:: A link:https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Driver[`Driver`] object holds the details required to establish connections with a Neo4j database. 
Every Neo4j-backed application requires a `Driver` object.
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[[LTS]]LTS:: A _Long Term Support_ release is one guaranteed to be supported for a number of years. Neo4j 4.4 is LTS, and Neo4j 5 will also have an LTS version.
+[[Aura]]Aura:: link:https://neo4j.com/cloud/platform/aura-graph-database/[Aura] is Neo4j's fully managed cloud service. It comes with both free and paid plans.
+[[Driver]]Driver:: A link:https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Driver[`Driver`] object holds the details required to establish connections with a Neo4j database. Every Neo4j-backed application requires a `Driver` object.
+[[Cypher]]Cypher:: link:https://neo4j.com/docs/getting-started/current/cypher-intro/[Cypher] is Neo4j's graph query language that lets you retrieve data from the graph. It is like SQL, but for graphs.
```
```suggestion
[[Cypher]]Cypher:: link:https://neo4j.com/docs/getting-started/current/cypher-intro/[Cypher] is Neo4j's graph query language that lets you retrieve data from the graph. 
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[[Driver]]Driver:: A link:https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Driver[`Driver`] object holds the details required to establish connections with a Neo4j database. Every Neo4j-backed application requires a `Driver` object.
+[[Cypher]]Cypher:: link:https://neo4j.com/docs/getting-started/current/cypher-intro/[Cypher] is Neo4j's graph query language that lets you retrieve data from the graph. It is like SQL, but for graphs.
+[[APOC]]APOC:: link:https://neo4j.com/labs/apoc/current/introduction/[Awesome Procedures On Cypher (APOC)] is a library of (many) functions that can not be easily expressed in Cypher itself.
+[[Bolt]]Bolt:: link:https://neo4j.com/docs/bolt/current/bolt/[Bolt] is the protocol used for interaction between Neo4j instances and drivers. It listens on port 7687 by default.
```
```suggestion
[[Bolt]]Bolt:: link:https://neo4j.com/docs/bolt/current/bolt/[Bolt] is the protocol used for interaction between Neo4j instances and drivers. 
It listens on port 7687 by default.
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[[Cypher]]Cypher:: link:https://neo4j.com/docs/getting-started/current/cypher-intro/[Cypher] is Neo4j's graph query language that lets you retrieve data from the graph. It is like SQL, but for graphs.
+[[APOC]]APOC:: link:https://neo4j.com/labs/apoc/current/introduction/[Awesome Procedures On Cypher (APOC)] is a library of (many) functions that can not be easily expressed in Cypher itself.
+[[Bolt]]Bolt:: link:https://neo4j.com/docs/bolt/current/bolt/[Bolt] is the protocol used for interaction between Neo4j instances and drivers. It listens on port 7687 by default.
+[[ACID]]ACID:: Atomicity, Consistency, Isolation, Durability (ACID) are properties guaranteeing that database transactions are processed reliably. An ACID-compliant DBMS ensures that the data in the database remains accurate and consistent despite failures.
```
```suggestion
[[ACID]]ACID:: Atomicity, Consistency, Isolation, Durability (ACID) are properties guaranteeing that database transactions are processed reliably. 
An ACID-compliant DBMS ensures that the data in the database remains accurate and consistent despite failures.
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[[APOC]]APOC:: link:https://neo4j.com/labs/apoc/current/introduction/[Awesome Procedures On Cypher (APOC)] is a library of (many) functions that can not be easily expressed in Cypher itself.
+[[Bolt]]Bolt:: link:https://neo4j.com/docs/bolt/current/bolt/[Bolt] is the protocol used for interaction between Neo4j instances and drivers. It listens on port 7687 by default.
+[[ACID]]ACID:: Atomicity, Consistency, Isolation, Durability (ACID) are properties guaranteeing that database transactions are processed reliably. An ACID-compliant DBMS ensures that the data in the database remains accurate and consistent despite failures.
+[[eventual_consistency]]eventual consistency:: A database is eventually consistent if it provides the guarantee that all cluster replicas will, _at some point in time_, store the latest version of the data.
```
We don't really refer to cluster members as replicas anymore, not since the introduction of autonomous clusters. A member of a cluster can simultaneously be a primary host for one db and a secondary host for another db. Referring to replicas of a cluster is misleading.

nvitucci reviewed (approved):
Looks really good now!

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[[Bolt]]Bolt:: link:https://neo4j.com/docs/bolt/current/bolt/[Bolt] is the protocol used for interaction between Neo4j instances and drivers. It listens on port 7687 by default.
+[[ACID]]ACID:: Atomicity, Consistency, Isolation, Durability (ACID) are properties guaranteeing that database transactions are processed reliably. An ACID-compliant DBMS ensures that the data in the database remains accurate and consistent despite failures.
+[[eventual_consistency]]eventual consistency:: A database is eventually consistent if it provides the guarantee that all cluster replicas will, _at some point in time_, store the latest version of the data.
+[[causal_consistency]]causal consistency:: A database is causally consistent if read and write queries are seen by every replica of the cluster in the same order. This is stronger than eventual consistency.
```
As above. 
```suggestion
[[causal_consistency]]causal consistency:: A database is causally consistent if read and write queries are seen by every member of the cluster in the same order. 
This is stronger than eventual consistency.
```

AlexicaWright commented on common-content/modules/ROOT/partials/glossary.adoc:
```diff
@@ -0,0 +1,13 @@
+[[causal_consistency]]causal consistency:: A database is causally consistent if read and write queries are seen by every replica of the cluster in the same order. This is stronger than eventual consistency.
+[[null]]null:: The null marker is not a type but a placeholder for absence of value.
+For more information, see link:/docs/cypher-manual/4.4/syntax/working-with-null[Cypher Manual -- Working with `null`].
+[[transaction]]transaction:: A transaction is a unit of work that is either _committed_ in its entirety or _rolled back_ on failure. An example is a bank transfer: it involves multiple steps, but they must _all_ succeed or be reverted, to avoid money being subtracted from one account but not added to the other.
```
```suggestion
[[transaction]]transaction:: A transaction is a unit of work that is either _committed_ in its entirety or _rolled back_ on failure. 
An example is a bank transfer: it involves multiple steps, but they must _all_ succeed or be reverted, to avoid money being subtracted from one account but not added to the other.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,103 @@
+= Coordinate parallel transactions
+
+When working with a Neo4j cluster, the driver automatically enforces <<causal_consistency>> for transactions within the same session, which guarantees that a query is able to read changes made by previous queries. The same does not happen by default for multiple xref:transactions.adoc[transactions] running in parallel though. In that case, you can use _bookmarks_ to have one transaction wait for  ...
```
```suggestion
When working with a Neo4j cluster, the driver automatically enforces <<causal_consistency>> for transactions within the same session, which guarantees that a query is able to read changes made by previous queries. 
The same does not happen by default for multiple xref:transactions.adoc[transactions] running in parallel though. 
In that case, you can use _bookmarks_ to have one transaction wait for the result of another to be propagated across the cluster before running its own work. 
This is not a requirement, and *you should only use bookmarks if you _need_ casual consistency across different transactions*.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,103 @@
+
+When working with a Neo4j cluster, the driver automatically enforces <<causal_consistency>> for transactions within the same session, which guarantees that a query is able to read changes made by previous queries. The same does not happen by default for multiple xref:transactions.adoc[transactions] running in parallel though. In that case, you can use _bookmarks_ to have one transaction wait for  ...
+
+A _bookmark_ is a token that represents some state of the database. By passing one or multiple bookmarks along with a query, the server will make sure that the query does not get executed before the represented state(s) have been established.
```
```suggestion
A _bookmark_ is a token that represents some state of the database. 
By passing one or multiple bookmarks along with a query, the serve makes sure that the query does not get executed before the represented state(s) has been established.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,103 @@
+
+== Bookmarks across multiple sessions
+
+If your application uses multiple sessions instead, you might need to ensure that one session has completed all its transactions before another session is allowed to run its queries. In those cases, you can collect the bookmarks from some sessions using the method `Session.last_bookmarks()` *(1), (2)*, store them into a `Bookmarks` object, and use them to initialize another session with the `book ...
```
```suggestion
If your application uses multiple sessions instead, you may need to ensure that one session has completed all its transactions before another session is allowed to run its queries. 
In those cases, you can collect the bookmarks from some sessions using the method `Session.last_bookmarks()` *(1), (2)*, store them into a `Bookmarks` object, and use them to initialize another session with the `bookmarks` parameter *(3)*.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,103 @@
+
+If your application uses multiple sessions instead, you might need to ensure that one session has completed all its transactions before another session is allowed to run its queries. In those cases, you can collect the bookmarks from some sessions using the method `Session.last_bookmarks()` *(1), (2)*, store them into a `Bookmarks` object, and use them to initialize another session with the `book ...
+
+In the example below, `session_a` and `session_b` are allowed to run concurrently, while `session_c` will wait until their results have been propagated. This guarantees the `Person` nodes `session_c` wants to act on actually exist.
```
```suggestion
In the example below, `session_a` and `session_b` are allowed to run concurrently, while `session_c` waits until their results have been propagated. 
This guarantees that the `Person` nodes `session_c` wants to act on actually exist.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,103 @@
+
+image:{common-image}/driver-passing-bookmarks.svg[]
+
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/bookmarks.adoc:
```diff
@@ -0,0 +1,103 @@
+image:{common-image}/driver-passing-bookmarks.svg[]
+
+[TIP]
+The usage of bookmarks can negatively impact performance, since all queries are forced to wait for the latest changes to be propagated across the cluster. For simple use-cases, try to group queries within a single transaction, or within a single session.
```
```suggestion
The use of bookmarks can negatively impact performance, since all queries are forced to wait for the latest changes to be propagated across the cluster. 
For simple use-cases, try to group queries within a single transaction, or within a single session.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/concurrency.adoc:
```diff
@@ -0,0 +1,42 @@
+= Run concurrent transactions
+
+The driver is compatible with Python's `asyncio`, which allows to implement concurrent workflows. To interact with the database in an asynchronous way, create an `AsyncDriver` with `AsyncGraphDatabase.driver()`. The workflow is very similar to the synchronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited. If you need <<cas ...
```
```suggestion
The driver is compatible with Python's `asyncio`, which allows implementing concurrent workflows. 
To interact with the database in an asynchronous way, create an `AsyncDriver` with `AsyncGraphDatabase.driver()`. 
The workflow is very similar to the synchronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited. 
If you need <<casual_consistency>> across different transactions, use xref:bookmarks.adoc[bookmarks].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/concurrency.adoc:
```diff
@@ -0,0 +1,42 @@
+asyncio.run(main())
+----
+
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/concurrency.adoc:
```diff
@@ -0,0 +1,42 @@
+----
+
+[TIP]
+Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although `asyncio` is often easier to implement in an application.
```
```suggestion
Async implements a concurrency model, but it is not the only possible one. 
Multithreading is also possible, although `asyncio` is often easier to implement in an application.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/concurrency.adoc:
```diff
@@ -0,0 +1,42 @@
+[TIP]
+Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although `asyncio` is often easier to implement in an application.
+
+[WARNING]
```
```suggestion
[WARNING]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/concurrency.adoc:
```diff
@@ -0,0 +1,42 @@
+Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although `asyncio` is often easier to implement in an application.
+
+[WARNING]
+There is a known issue with Python 3.8 and the async driver where it gradually slows down. It is generally recommended to use the latest supported version of Python for the best performance, stability, and security.
```
```suggestion
There is a known issue with Python 3.8 and the async driver where it gradually slows down. 
It is generally recommended to use the latest supported version of Python for the best performance, stability, and security.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+
+== Connection protocols and security
+
+Communication between the driver and the server is mediated by <<Bolt>>. The scheme of the server URI determines whether the connection is encrypted and, if so, what type of certificates are accepted.
```
```suggestion
Communication between the driver and the server is mediated by <<Bolt>>. 
The scheme of the server URI determines whether the connection is encrypted and, if so, what type of certificates are accepted.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+|
+|===
+
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+|===
+
+[TIP]
+The driver receives a _routing table_ from the server upon successful connection, regardless of whether the instance is a proper cluster environment or a single-machine environment. The driver's routing behavior works in tandem with link:{neo4j-docs-base-uri}/operations-manual/current/clustering/[Neo4j's clustering] by directing read/write transactions to appropriate cluster members. If you want  ...
```
```suggestion
The driver receives a _routing table_ from the server upon successful connection, regardless of whether the instance is a proper cluster environment or a single-machine environment. 
The driver's routing behavior works in tandem with link:{neo4j-docs-base-uri}/operations-manual/current/clustering/[Neo4j's clustering] by directing read/write transactions to appropriate cluster members. 
If you want to target a _specific_ machine, use a link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#bolt-driver-ref[`BoltDriver`] object and `bolt`, `bolt+s`, or `bolt+ssc` URI schemes instead.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+
+=== Basic authentication (default)
+
+The basic authentication scheme relies on traditional username and password. These can either be the credentials for your local installation, or the ones provided with an Aura instance.
```
```suggestion
The basic authentication scheme relies on traditional username and password. 
These can either be the credentials for your local installation, or the ones provided with an Aura instance.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+driver = GraphDatabase.driver(uri, auth=(user, password))
+----
+
+[INFO]
```
```suggestion
[INFO]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+----
+
+[INFO]
+The basic authentication scheme can also be used to authenticate against an LDAP server (Enterprise Edition only).
```
```suggestion
The basic authentication scheme can also be used to authenticate against an LDAP server (Enterprise Edition only).
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+driver = GraphDatabase.driver(uri, auth=bearer_auth(token))
+----
+
+[NOTE]
```
```suggestion
[NOTE]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+----
+
+[NOTE]
+The bearer authentication scheme requires link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/sso-integration[configuring Single Sign-On on the server]. Once configured, clients can discover Neo4j's configuration through the link:{neo4j-docs-base-uri}/http-api/current/discovery/[Discovery API].
```
```suggestion
The bearer authentication scheme requires link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/sso-integration[configuring Single Sign-On on the server]. 
Once configured, clients can discover Neo4j's configuration through the link:{neo4j-docs-base-uri}/http-api/current/discovery/[Discovery API].
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+
+Use the function link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.custom_auth[`custom_auth`] to log into a server having a custom authentication scheme.
+
+[NOTE]
```
```suggestion
[NOTE]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+Use the function link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.custom_auth[`custom_auth`] to log into a server having a custom authentication scheme.
+
+[NOTE]
+If authentication is disabled on the server, the `auth` parameter can be omitted entirely.
```
```suggestion
If authentication is disabled on the server, the `auth` parameter can be omitted entirely.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect-advanced.adoc:
```diff
@@ -0,0 +1,117 @@
+
+== Custom address resolver
+
+You can specify a custom _resolver_ function to resolve host and port values ahead of DNS resolution. This function is called with a 2-tuple of `(host, port)` and should return an iterable of 2-tuples `(host, port)`.
```
```suggestion
You can specify a custom _resolver_ function to resolve host and port values ahead of DNS resolution. 
This function is called with a 2-tuple of `(host, port)` and should return an iterable of 2-tuples `(host, port)`.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect.adoc:
```diff
@@ -0,0 +1,43 @@
+    driver.verify_connectivity() # <2>
+----
+
+Creating a `Driver` instance *(1)* only provides information on _how_ to access the database, but does not actually _establish_ a connection. Connection is instead deferred to when the first query is executed. To verify immediately that the driver can connect to the database (valid credentials, compatible versions, etc), use the `.verify_connectivity()` method *(2)* after initializing the driver.
```
```suggestion
Creating a `Driver` instance *(1)* only provides information on _how_ to access the database, but does not actually _establish_ a connection. 
Connection is instead deferred to when the first query is executed. 
To verify immediately that the driver can connect to the database (valid credentials, compatible versions, etc), use the `.verify_connectivity()` method *(2)* after initializing the driver.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect.adoc:
```diff
@@ -0,0 +1,43 @@
+
+Creating a `Driver` instance *(1)* only provides information on _how_ to access the database, but does not actually _establish_ a connection. Connection is instead deferred to when the first query is executed. To verify immediately that the driver can connect to the database (valid credentials, compatible versions, etc), use the `.verify_connectivity()` method *(2)* after initializing the driver.
+
+Both the creation of a `Driver` object and the connection verification can raise a number of different link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#errors[exceptions]. Since error handling can get quite verbose, and a connection error is a blocker for any subsequnt task, the most common choice is to let the program crash should an exception occurr during connection.
```
```suggestion
Both the creation of a `Driver` object and the connection verification can raise a number of different link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#errors[exceptions]. 
Since error handling can get quite verbose, and a connection error is a blocker for any subsequent task, the most common solution is to let the program crash if an exception occurs during connection.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect.adoc:
```diff
@@ -0,0 +1,43 @@
+Both the creation of a `Driver` object and the connection verification can raise a number of different link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#errors[exceptions]. Since error handling can get quite verbose, and a connection error is a blocker for any subsequnt task, the most common choice is to let the program crash should an exception occurr during connection.
+
+`Driver` objects are _immutable_, _thread-safe_, and fairly _expensive to create_.
+Share them across threads (but not across processes!) and use xref:transactions#impersonation[impersonation] to query the database with a different user.
```
```suggestion
Share them across threads (but not across processes) and use xref:transactions#impersonation[impersonation] to query the database with a different user.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/connect.adoc:
```diff
@@ -0,0 +1,43 @@
+
+== Close connections
+
+Always close `Driver` objects to free up all allocated resources, even upon unsuccessful connection or runtime errors. Either instantiate the `Driver` object using the `with` statement, or call the `Driver.close()` method explicitly.
```
```suggestion
Always close `Driver` objects to free up all allocated resources, even upon unsuccessful connection or runtime errors. 
Either instantiate the `Driver` object using the `with` statement, or call the `Driver.close()` method explicitly.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/index.adoc:
```diff
@@ -1,51 +1,86 @@
 
-License: <<license, Creative Commons 4.0>>
-endif::[]
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/index.adoc:
```diff
@@ -1,51 +1,86 @@
-License: <<license, Creative Commons 4.0>>
-endif::[]
+[TIP]
+You need a running Neo4j database to use the driver with it. If you do not have one yet, xref:install#get-an-instance[get an instance] for free.
```
```suggestion
You need a running Neo4j database to use the driver with it.
 If you do not have one yet, xref:install#get-an-instance[get an instance] for free.
 ====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/index.adoc:
```diff
@@ -1,51 +1,86 @@
 
-[.discrete]
-=== link:https://graphacademy.neo4j.com/courses/app-python/?ref=guides[Building Neo4j Applications with Python^]
+Connect to a database by creating a <<Driver>> object and providing a URL and an authentication token. Once you have a `Driver` instance, use the `.verify_connectivity()` method to ensure that a working connection can be established.
```
```suggestion
Connect to a database by creating a <<Driver>> object and providing a URL and an authentication token. 
Once you have a `Driver` instance, use the `.verify_connectivity()` method to ensure that a working connection can be established.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/install.adoc:
```diff
@@ -0,0 +1,47 @@
+pip install neo4j
+----
+
+Always use the latest version of the driver, as it will always work both with the previous Neo4j <<LTS>> release and with the current and next major releases.
```
```suggestion
Always use the latest version of the driver, as it always works both with the previous Neo4j <<LTS>> release and with the current and next major releases.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/install.adoc:
```diff
@@ -0,0 +1,47 @@
+----
+
+Always use the latest version of the driver, as it will always work both with the previous Neo4j <<LTS>> release and with the current and next major releases.
+The latest `5.x` driver supports connection to any Neo4j 5 and 4.4 instance, and will also be compatible with Neo4j 6. For a detailed list of changes across versions, see the link:https://github.com/neo4j/neo4j-python-driver/wiki/5.x-changelog[driver's changelog].
```
```suggestion
The latest `5.x` driver supports connection to any Neo4j 5 and 4.4 instances, and is also compatible with Neo4j 6. For a detailed list of changes across versions, see the link:https://github.com/neo4j/neo4j-python-driver/wiki/5.x-changelog[driver's changelog].
```
Neo4j 6? It's not even on the horizon yet.

stefano-ottolenghi replied:
Forward and backward compatibility is important for developers, to have the guarantee that whatever work they invest today will be valid as long as possible. The Drivers team makes this commitment by guaranteeing that if you invest in developing with the driver v5, you can keep your code as-is for not only v5, but also for the next major. The same has happened with driver v4: it works with Neo 5.

AlexicaWright commented on python-manual/modules/ROOT/pages/install.adoc:
```diff
@@ -0,0 +1,47 @@
+Always use the latest version of the driver, as it will always work both with the previous Neo4j <<LTS>> release and with the current and next major releases.
+The latest `5.x` driver supports connection to any Neo4j 5 and 4.4 instance, and will also be compatible with Neo4j 6. For a detailed list of changes across versions, see the link:https://github.com/neo4j/neo4j-python-driver/wiki/5.x-changelog[driver's changelog].
+
+[NOTE]
```
```suggestion
[NOTE]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/install.adoc:
```diff
@@ -0,0 +1,47 @@
+The latest `5.x` driver supports connection to any Neo4j 5 and 4.4 instance, and will also be compatible with Neo4j 6. For a detailed list of changes across versions, see the link:https://github.com/neo4j/neo4j-python-driver/wiki/5.x-changelog[driver's changelog].
+
+[NOTE]
+To get the driver on an air-gapped machine, link:https://pypi.org/project/neo4j/#files[download the latest driver] tarball and install it with `pip install neo4j-<version>.tar.gz`.
```
```suggestion
To get the driver on an air-gapped machine, link:https://pypi.org/project/neo4j/#files[download the latest driver] tarball and install it with `pip install neo4j-<version>.tar.gz`.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/install.adoc:
```diff
@@ -0,0 +1,47 @@
+== Get a Neo4j instance
+
+You need a running Neo4j database in order to use the driver with it.
+The easiest way to spin up a *local instance* is through a link:{neo4j-docs-base-uri}/operations-manual/current/docker/introduction/[Docker container] (requires `docker.io`). The command below runs the latest Neo4j version in Docker, setting the admin username to `neo4j` and password to `secretgraph`:
```
```suggestion
The easiest way to spin up a *local instance* is through a link:{neo4j-docs-base-uri}/operations-manual/current/docker/introduction/[Docker container] (requires `docker.io`). 
The command below runs the latest Neo4j version in Docker, setting the admin username to `neo4j` and password to `secretgraph`:
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+= Performance recommendations
+
+- *Specify the target database on all queries* with the `database` parameter when creating new xref:transactions.adoc[sessions]. If no database is provided, the driver has to send an extra request to the server to figure out what the default database is. The overhead is minimal for a single session, but becomes significant over hundreds of sessions.
```
```suggestion
- *Specify the target database on all queries* with the `database` parameter when creating new xref:transactions.adoc[sessions]. 
If no database is provided, the driver has to send an extra request to the server to figure out what the default database is. 
The overhead is minimal for a single session, but becomes significant over hundreds of sessions.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+
+// - Execute as many queries within the same session. Minimise the number of transactions?
+
+- *Use xref:query-simple#query-parameters[query parameters]* instead of hardcoding or concatenating values into queries. This allows the database to properly cache queries.
```
```suggestion
- *Use xref:query-simple#query-parameters[query parameters]* instead of hardcoding or concatenating values into queries. 
This allows the database to properly cache queries.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+session.run("MATCH (p:Person {name: " + name + "}) RETURN p")
+----
+
+- *Specify node labels* in all queries. To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cypher -- Label expressions].
```
```suggestion
- *Specify node labels* in all queries. 
To learn how to combine labels, see link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/expressions/#query-syntax-label[Cypher -- Label expressions].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+    session.run("MERGE (:Number {value: $value})", value=random())
+----
++
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
++
+[TIP]
+The most efficient way of performing a _first import_ of large amounts of data into a new database is the link:{neo4j-docs-base-uri}/operations-manual/current/tutorial/neo4j-admin-import/[`neo4j-admin database import`] command.
+// source:  https://community.neo4j.com/t5/neo4j-graph-platform/improving-data-writing-efficiency-in-python/td-p/39520
```
```suggestion
// source:  https://community.neo4j.com/t5/neo4j-graph-platform/improving-data-writing-efficiency-in-python/td-p/39520
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+session.run("MATCH (p:Person) WHERE p.name = $name RETURN p", name="Alice")
+----
+
+- *Create indexes for properties that you often filter against*. For example, if you often look up `Person` nodes by the `name` property, it is beneficial to create an index on `Person.name`. You can create indexes with the `CREATE INDEX` Cypher function, for both nodes and relationships. For more information, see link:{neo4j-docs-base-uri}/cypher-manual/current/indexes-for-search-performance/[In ...
```
```suggestion
- *Create indexes for properties that you often filter against*. 
For example, if you often look up `Person` nodes by the `name` property, it is beneficial to create an index on `Person.name`. 
You can create indexes with the `CREATE INDEX` Cypher function, for both nodes and relationships. 
For more information, see link:{neo4j-docs-base-uri}/cypher-manual/current/indexes-for-search-performance/[Indexes for search performance].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+session.run("CREATE INDEX person_name FOR (n:Person) ON (n.name)")
+----
+
+- link:{neo4j-docs-base-uri}/cypher-manual/current/query-tuning/basic-example/#_profile_query[*Profile your queries*] to locate queries whose performance can be improved. You can profile queries by prepending them with `PROFILE`. The server output is available in the `profile` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.
```
```suggestion
- link:{neo4j-docs-base-uri}/cypher-manual/current/query-tuning/basic-example/#_profile_query[*Profile your queries*] to locate queries whose performance can be improved. 
You can profile queries by prepending them with `PROFILE`. 
The server output is available in the `profile` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+"""
+----
++
+In case some queries were so slow that you are unable to even run them in reasonable times, you can prepend them with `EXPLAIN` instead of `PROFILE`. This will return the _plan_ that the server would use to run the query, but without executing it. The server output is available in the `plan` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary ...
```
```suggestion
In case some queries are so slow that you are unable to even run them in reasonable times, you can prepend them with `EXPLAIN` instead of `PROFILE`. 
This returns the _plan_ that the server would use to run the query, but without executing it. 
The server output is available in the `plan` property of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`] object.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/performance.adoc:
```diff
@@ -0,0 +1,160 @@
+"""
+----
+
+- *Use xref:concurrency.adoc[concurrency]*, either in the form of multithreading or with the async version of the driver. This is likely to be more impactful on performance if you parallelize complex and time-consuming queries in your application, but not so much if you run many simple ones.
```
```suggestion
- *Use xref:concurrency.adoc[concurrency]*, either in the form of multithreading or with the async version of the driver. 
This is likely to be more impactful on performance if you parallelize complex and time-consuming queries in your application, but not so much if you run many simple ones.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+
+== Auto-commit transactions
+
+This is the most basic and limited form with which to run a Cypher query. The driver will not automatically retry auto-commit transactions, as it does instead for queries run with xref:transactions#managed-transactions[managed transactions].
```
```suggestion
This is the most basic and limited form to run a Cypher query wtih. 
The driver does not automatically retry auto-commit transactions, as it does for queries run with xref:transactions#managed-transactions[managed transactions].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+
+This is the most basic and limited form with which to run a Cypher query. The driver will not automatically retry auto-commit transactions, as it does instead for queries run with xref:transactions#managed-transactions[managed transactions].
+
+You run an auto-commit transaction with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.run[`Session.run()`]. It returns a `Result` object that needs to be processed with the xref:result#manipulate-result[related methods].
```
```suggestion
You run an auto-commit transaction with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.run[`Session.run()`]. 
It returns a `Result` object that needs to be processed with the xref:result#manipulate-result[related methods].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+    session.run("CREATE (a:Person {name: $name})", name=name)
+----
+
+An auto-commit transaction will be committed _at the latest_ when the session is destroyed, or before another transaction is executed within the same session. Other than that, there is no clear guarantee on when exactly an auto-commit transaction will be committed during the lifetime of a session. To ensure an auto-commit transaction is committed, you can call the `.consume()` method on its resul ...
```
```suggestion
An auto-commit transaction is committed _at the latest_ when the session is destroyed, or before another transaction is executed within the same session. 
Other than that, there is no clear guarantee of when exactly an auto-commit transaction is committed during the lifetime of a session. 
To ensure an auto-commit transaction is committed, you can call the `.consume()` method on its result.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+
+Auto-commit transactions should only be used when xref:transactions#managed-transactions[transaction functions] do not fit the purpose, or for quick prototyping.
+
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+Auto-commit transactions should only be used when xref:transactions#managed-transactions[transaction functions] do not fit the purpose, or for quick prototyping.
+
+[TIP]
+*This method is the only one that can be used for link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/call-subquery/#subquery-call-in-transactions[`CALL {} IN TRANSACTIONS`] and link:{neo4j-docs-base-uri}/cypher-manual/current/query-tuning/using/#query-using-periodic-commit-hint[`PERIODIC COMMIT`] queries*.
```
```suggestion
*This method is the only one that can be used for link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/call-subquery/#subquery-call-in-transactions[`CALL {} IN TRANSACTIONS`] and link:{neo4j-docs-base-uri}/cypher-manual/current/query-tuning/using/#query-using-periodic-commit-hint[`PERIODIC COMMIT`] queries*.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+
+=== Configure auto-commit transactions with the `Query` object
+
+The link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Query[`Query`] object allows to specify a query timeout and to attach metadata to the transaction, which will be visible on the server (as described for the xref:transactions#unit_of_work-decorator[`unit_of_work`] decorator).
```
```suggestion
The link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Query[`Query`] object allows to specify a query timeout and to attach metadata to the transaction, which is visible on the server (as described for the xref:transactions#unit_of_work-decorator[`unit_of_work`] decorator).
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+
+== Dynamic values in property keys, relationship types, and labels
+
+In general, you should not concatenate parameters directly into a query, but rather use xref:query-simple#query-parameters[query parameters]. There can however be circumstances where your query structure prevents the usage of parameters in all its parts. In fact, although parameters can be used for literals and expressions as well as node and relationship ids, they cannot be used for the followin ...
```
```suggestion
In general, you should not concatenate parameters directly into a query, but rather use xref:query-simple#query-parameters[query parameters]. 
There can, however, be circumstances where your query structure prevents the usage of parameters in all its parts. 
In fact, although parameters can be used for literals and expressions as well as node and relationship ids, they cannot be used for the following constructs:
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+- relationship types, so `MATCH (n)-[:$param]->(m)` is invalid;
+- labels, so `MATCH (n:$param)` is invalid.
+
+For those queries, you are forced to use string concatenation. To protect against Cypher injections you should enclose the dynamic values in backticks and escape them yourself. Notice that Cypher processes Unicode, so take care of the Unicode literal `\u0060` as well.
```
```suggestion
For those queries, you are forced to use string concatenation. 
To protect against Cypher injections you should enclose the dynamic values in backticks and escape them yourself.
Note that Cypher processes Unicode, so enclose the Unicode literal `\u0060` as well.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+    )
+----
+
+Another workaround, which avoids string concatenation, is using the <<APOC>> procedure link:{neo4j-docs-base-uri}/apoc/current/overview/apoc.merge/apoc.merge.node/[`apoc.merge.node`]. It supports dynamic labels and property keys, but only for node insertion.
```
```suggestion
Another workaround, which avoids string concatenation, is using the <<APOC>> procedure link:{neo4j-docs-base-uri}/apoc/current/overview/apoc.merge/apoc.merge.node/[`apoc.merge.node`]. 
It supports dynamic labels and property keys, but only for node insertion.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+    )
+----
+
+[NOTE]
```
```suggestion
[NOTE]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+----
+
+[NOTE]
+If you are running Neo4j in Docker, APOC needs to be enabled when starting the container. See link:https://neo4j.com/labs/apoc/current/installation/#docker[APOC - Installation - Docker].
```
```suggestion
If you are running Neo4j in Docker, APOC needs to be enabled when starting the container. See link:https://neo4j.com/labs/apoc/current/installation/#docker[APOC - Installation - Docker].
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-advanced.adoc:
```diff
@@ -0,0 +1,114 @@
+
+== Logging
+
+The driver logs messages through the native `logging` library to a logger named `neo4j`. To redirect log messages to standard output, use the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.debug.watch[watch] function:
```
```suggestion
The driver logs messages through the native `logging` library to a logger named `neo4j`. 
To redirect log messages to standard output, use the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.debug.watch[watch] function:
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+    ))
+----
+
+where *(1)* specifies the Cypher query and *(2)* is a _query parameter_. The query and its parameters are wrapped in a transaction function *(3)*, which is passed as callback to `Session.execute_write()` together with an arbitrary numbers of keyword arguments *(4)*.
```
```suggestion
where *(1)* specifies the Cypher query and *(2)* is a _query parameter_. 
The query and its parameters are wrapped in a transaction function *(3)*, which is passed as callback to `Session.execute_write()` together with an arbitrary numbers of keyword arguments *(4)*.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+
+where *(1)* specifies the Cypher query and *(2)* is a _query parameter_. The query and its parameters are wrapped in a transaction function *(3)*, which is passed as callback to `Session.execute_write()` together with an arbitrary numbers of keyword arguments *(4)*.
+
+[NOTE]
```
```suggestion
[NOTE]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+where *(1)* specifies the Cypher query and *(2)* is a _query parameter_. The query and its parameters are wrapped in a transaction function *(3)*, which is passed as callback to `Session.execute_write()` together with an arbitrary numbers of keyword arguments *(4)*.
+
+[NOTE]
+`MERGE` creates a new node matching the requirements unless one already exists, in which case nothing is done. For strict node creation, use the link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/create/[`CREATE`] clause.
```
```suggestion
`MERGE` creates a new node matching the requirements unless one already exists, in which case nothing is done. 
For strict node creation, use the link:{neo4j-docs-base-uri}/cypher-manual/current/clauses/create/[`CREATE`] clause.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+====
+It might feel tempting to create new relationships with a single `MERGE` clause, such as: +
+`MERGE (:Person {name: "Alice"})-[KNOWS]->(:Person {name: "Bob"})`. +
+However, this would result in the creation of an _extra_ `Alice` node, so that you would end up with unintended duplicate records. To avoid this, always `MATCH` the elements that you want to update, and use the resulting reference in the `MERGE` clause (as shown in the previous example). See link:https://neo4j.com/developer/kb/understanding-how-merge-works/[Understanding how MERGE works].
```
```suggestion
However, this would result in the creation of an _extra_ `Alice` node, so that you would end up with unintended duplicate records. 
To avoid this, always `MATCH` the elements that you want to update, and use the resulting reference in the `MERGE` clause (as shown in the previous example). 
See link:https://neo4j.com/developer/kb/understanding-how-merge-works/[Understanding how MERGE works].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+[#query-parameters]
+== How to pass parameters to queries
+
+*Do not hardcode or concatenate parameters directly into queries*. Instead, always use placeholders and specify the link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/parameters/[Cypher parameters] as keyword arguments or in a dictionary, as shown in the previous examples. This is for:
```
```suggestion
*Do not hardcode or concatenate parameters directly into queries*. 
Instead, always use placeholders and specify the link:{neo4j-docs-base-uri}/cypher-manual/current/syntax/parameters/[Cypher parameters] as keyword arguments or in a dictionary, as shown in the previous examples. 
This is for:
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+1. *performance benefits*: Neo4j compiles and caches queries, but can only do so if the query structure is unchanged;
+2. *security reasons*: link:https://neo4j.com/developer/kb/protecting-against-cypher-injection/[protecting against Cypher injection].
+
+[NOTE]
```
```suggestion
[NOTE]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+2. *security reasons*: link:https://neo4j.com/developer/kb/protecting-against-cypher-injection/[protecting against Cypher injection].
+
+[NOTE]
+There can be circumstances where your query structure prevents the usage of parameters in all its parts. For those advanced use cases, see xref:query-advanced#_dynamic_values_in_property_keys_relationship_types_and_labels[Dynamic values in property keys, relationship types, and labels].
```
```suggestion
There can be circumstances where your query structure prevents the usage of parameters in all its parts. 
For those advanced use cases, see xref:query-advanced#_dynamic_values_in_property_keys_relationship_types_and_labels[Dynamic values in property keys, relationship types, and labels].
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+
+== Anatomy of a query
+
+Before you can run a query, you need to obtain a session from the driver with the `Driver.session()` method *(1)*. The `database` parameter is optional but recommended for performance, and further xref:transactions#_session_configuration[configuration parameters] can be included.
```
```suggestion
Before you can run a query, you need to obtain a session from the driver with the `Driver.session()` method *(1)*. 
The `database` parameter is optional but recommended for performance, and further xref:transactions#_session_configuration[configuration parameters] can be included.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+
+Before you can run a query, you need to obtain a session from the driver with the `Driver.session()` method *(1)*. The `database` parameter is optional but recommended for performance, and further xref:transactions#_session_configuration[configuration parameters] can be included.
+
+Within a session, use the methods link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_read[`Session.execute_read()`] and link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_write[`Session.execute_write()`] *(2)*, depending on whether you want to retrieve data from the database or alter it. Both methods take a _transaction function_ c ...
```
```suggestion
Within a session, use the methods link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_read[`Session.execute_read()`] and link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_write[`Session.execute_write()`] *(2)*, depending on whether you want to retrieve data from the database or alter it. 
Both methods take a _transaction function_ callback *(3)* and an arbitrary number of positional and keyword arguments *(4)* which are handed down to the transaction function.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+
+Within a session, use the methods link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_read[`Session.execute_read()`] and link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_write[`Session.execute_write()`] *(2)*, depending on whether you want to retrieve data from the database or alter it. Both methods take a _transaction function_ c ...
+
+The transaction function *(5)* is responsible for actually carrying out the queries and xref:result.adoc[processing the result]. Queries are specified with the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Transaction.run[`Transaction.run()`] method *(6)*, which returns a link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Result[`Result`] object. You can the ...
```
```suggestion
The transaction function *(5)* is responsible for actually carrying out the queries and xref:result.adoc[processing the result]. 
Queries are specified with the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Transaction.run[`Transaction.run()`] method *(6)*, which returns a link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Result[`Result`] object. 
You can then process the result *(7)* using xref:result#manipulate-result[any of the `Result` methods], or simply casting it to `list`.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+
+You can find more information about sessions and transactions in the section xref:transactions.adoc[Run your own transactions].
+
+[TIP]
```
```suggestion
[TIP]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/query-simple.adoc:
```diff
@@ -0,0 +1,251 @@
+You can find more information about sessions and transactions in the section xref:transactions.adoc[Run your own transactions].
+
+[TIP]
+The driver might automatically retry to run a failed transaction. For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know how many times they might be executed. In practice, this means that you should not edit nor rely on globals, for example. Note that although transactions functions might be execut ...
```
```suggestion
The driver might automatically retry to run a failed transaction. 
For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know how many times they might be executed. 
In practice, this means that you should not edit nor rely on globals, for example. 
Note that although transaction functions might be executed multiple times, the queries inside it are always run only once.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+= Manipulate query results
+
+In the xref:query-simple.adoc[basic query workflows] examples, transaction functions never returned a query result directly. Instead, they always _manipulated_ it in some way: either casting it to list, or calling a specific method. This is because the driver's output of a query is a `Result` object, which does not directly contain the result records. Rather, it encapsulates the Cypher result in  ...
```
```suggestion
In the xref:query-simple.adoc[basic query workflows] examples, transaction functions never returned a query result directly. 
Instead, they always _manipulated_ it in some way: either casting it to list or calling a specific method. 
This is because the driver's output of a query is a `Result` object, which does not directly contain the result records. 
Rather, it encapsulates the Cypher result in a rich data structure that requires some parsing within the transaction function.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+[#manipulate-result]
+== Manipulate the result
+
+When running a query, *the result records are not immediately and entirely fetched and returned by the server*. Instead, results come as a _lazy stream_. In particular, when the driver receives some records from the server, they are initially _buffered_ in a background queue. Records stay in the buffer until they are _consumed_ by the application, at which point they are _removed from the buffer_ ...
```
```suggestion
When running a query, *the result records are not immediately and entirely fetched and returned by the server*. 
Instead, results come as a _lazy stream_. 
In particular, when the driver receives some records from the server, they are initially _buffered_ in a background queue. 
Records stay in the buffer until they are _consumed_ by the application, at which point they are _removed from the buffer_. 
There is thus no way to retrieve a previous record from the stream, unless manually saved in an auxiliary data structure. 
When no more records are available, the result is _exhausted_.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+
+When running a query, *the result records are not immediately and entirely fetched and returned by the server*. Instead, results come as a _lazy stream_. In particular, when the driver receives some records from the server, they are initially _buffered_ in a background queue. Records stay in the buffer until they are _consumed_ by the application, at which point they are _removed from the buffer_ ...
+
+It is up to the transaction function to do all the processing. To process a result, you have two main options:
```
```suggestion
It is up to the transaction function to do all the processing. 
To process a result, you have two main options:
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+
+It is up to the transaction function to do all the processing. To process a result, you have two main options:
+
+1. cast the result object to `list`, which yields a list of link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Record[`Record`] objects. Use the `.data()` method on a `Record` to obtain a dict-like view, or the `.get(key, default=None)` method to retrieve the value for a given key.
```
```suggestion
1. cast the result object to `list`, which yields a list of link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Record[`Record`] objects. 
Use the `.data()` method on a `Record` to obtain a dict-like view, or the `.get(key, default=None)` method to retrieve the value for a given key.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+        print(person.data())  # or person.get("name")
+----
+
+2. use any of the of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#result[`Result`] methods. The ones most commonly needed are listed in the table below.
```
```suggestion
2. use any of the of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#result[`Result`] methods. The ones most commonly needed are listed in the table below.
```
```suggestion
2. use any of the of the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#result[`Result`] methods. 
The ones most commonly needed are listed in the table below.
```
Does this render as expected?

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+|Name |Description
+
+|`value(key=0, default=None)`
+|Return the remainder of the result as a list. If `key` is specified, only the given property will be included, and `default` allows to specify a value for nodes lacking that property.
```
```suggestion
|Return the remainder of the result as a list. If `key` is specified, only the given property is included, and `default` allows to specify a value for nodes lacking that property.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+|Return up to `n` records from the result.
+
+|`single(strict=False)`
+|Return the next and only remaining record, or `None`. Calling this method always exhausts the result.
```
```suggestion
| Return the next and only remaining record, or `None`. 
Calling this method always exhausts the result.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+|Return the remainder of the result as a list. If `key` is specified, only the given property will be included, and `default` allows to specify a value for nodes lacking that property.
+
+|`fetch(n)`
+|Return up to `n` records from the result.
```
```suggestion
| Return up to `n` records from the result.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+- `strict==True` -- a `ResultNotSingleError` is raised.
+
+|`peek()`
+|Return the next record from the result without consuming it. This leaves the record in the buffer for further processing.
```
```suggestion
| Return the next record from the result without consuming it. 
This leaves the record in the buffer for further processing.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+|Return the next record from the result without consuming it. This leaves the record in the buffer for further processing.
+
+|`data(*keys)`
+|Return a JSON-like dump of the raw result. Only use it for debugging purposes.
```
```suggestion
| Return a JSON-like dump of the raw result. 
Only use it for debugging purposes.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+|Return a JSON-like dump of the raw result. Only use it for debugging purposes.
+
+|`consume()`
+|Return the query link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`]. It exhausts the result, so should only be called when data processing is over.
```
```suggestion
| Return the query link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#resultsummary[`ResultSummary`]. 
It exhausts the result, so should only be called when data processing is over.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
++
+See the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#result[API documentation] for a complete list of `Result` methods.
+
+[WARNING]
```
```suggestion
[WARNING]
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+See the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#result[API documentation] for a complete list of `Result` methods.
+
+[WARNING]
+A transaction function must not return the `neo4j.Result` itself. Doing so is roughly equivalent to returning a _pointer_ to the result buffer, which gets invalidated as soon as the query's transaction is over.
```
```suggestion
A transaction function must not return the `neo4j.Result` itself. 
Doing so is roughly equivalent to returning a _pointer_ to the result buffer, which gets invalidated as soon as the query's transaction is over.
====
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+
+== Transform to pandas DataFrame
+
+The method `.to_df()` converts the result into a pandas DataFrame. This method is only available if the `pandas` library is installed.
```
```suggestion
The method `.to_df()` converts the result into a pandas DataFrame. 
This method is only available if the `pandas` library is installed.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+
+This method accepts two optional arguments:
+
+- `expand` -- If `True`, some data structures in the result will be recursively expanded and flattened. More info in the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Result.to_df[API documentation].
```
```suggestion
- `expand` -- If `True`, some data structures in the result are recursively expanded and flattened. 
More info in the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Result.to_df[API documentation].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+This method accepts two optional arguments:
+
+- `expand` -- If `True`, some data structures in the result will be recursively expanded and flattened. More info in the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Result.to_df[API documentation].
+- `parse_dates` -- If `True`, columns exclusively containing `time.DateTime` objects, `time.Date` objects, or `None`, will be converted to `pandas.Timestamp`.
```
```suggestion
- `parse_dates` -- If `True`, columns exclusively containing `time.DateTime` objects, `time.Date` objects, or `None`, are converted to `pandas.Timestamp`.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+
+== Transform to graph
+
+The method `.graph()` converts the result into a graph. To make the most out of this method, your query should return a graph-like result instead of a single column. The resulting graph object exposes the properties `nodes` and `relationships`, which are set views into xref:data-types#node[`Node`] and xref:data-types#relationship[`Relationship`] objects.
```
```suggestion
The method `.graph()` converts the result into a graph. 
To make the most out of this method, your query should return a graph-like result instead of a single column. 
The resulting graph object exposes the properties `nodes` and `relationships`, which are set views into xref:data-types#node[`Node`] and xref:data-types#relationship[`Relationship`] objects.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/result.adoc:
```diff
@@ -0,0 +1,236 @@
+
+The method `.graph()` converts the result into a graph. To make the most out of this method, your query should return a graph-like result instead of a single column. The resulting graph object exposes the properties `nodes` and `relationships`, which are set views into xref:data-types#node[`Node`] and xref:data-types#relationship[`Relationship`] objects.
+
+You can use the graph format for further processing or to visualize the query result. An example implementation that uses the `pyvis` library to draw the graph is below.
```
```suggestion
You can use the graph format for further processing or to visualize the query result. 
An example implementation that uses the `pyvis` library to draw the graph is below.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+= Run your own transactions
+
+When xref:query-simple.adoc[querying the database] with transaction functions, the driver automatically creates a *<<transaction>>*. You can easily include multiple queries in a single transaction function, but to design more involved use-cases it is important to understand how transactions work within the driver.
```
```suggestion
When xref:query-simple.adoc[querying the database] with transaction functions, the driver automatically creates a *<<transaction>>*. 
You can easily include multiple queries in a single transaction function, but to design more involved use-cases it is important to understand how transactions work within the driver.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+== Obtain a session
+
+Before running a transaction, you need to obtain a _session_. Sessions act as concrete query channels between the driver and the server. In particular, sessions borrow connections from the _connection pool_ as needed, and ensure <<causal consistency>> is enforced.
```
```suggestion
Before running a transaction, you need to obtain a _session_. 
Sessions act as concrete query channels between the driver and the server. 
In particular, sessions borrow connections from the _connection pool_ as needed, and ensure <<causal consistency>> is enforced.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+Before running a transaction, you need to obtain a _session_. Sessions act as concrete query channels between the driver and the server. In particular, sessions borrow connections from the _connection pool_ as needed, and ensure <<causal consistency>> is enforced.
+
+Sessions are created with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#session[`Driver.session()`], with the keyword argument `database` allowing to specify the xref:_database_selection[target database]. For further parameters, see xref:_session_configuration[Session configuration].
```
```suggestion
Sessions are created with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#session[`Driver.session()`], with the keyword argument `database` allowing to specify the xref:_database_selection[target database]. For further parameters, see xref:_session_configuration[Session configuration].
```
```suggestion
Sessions are created with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#session[`Driver.session()`], with the keyword argument `database` allowing to specify the xref:_database_selection[target database]. 
For further parameters, see xref:_session_configuration[Session configuration].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+    ...
+----
+
+Session creation is a lightweight operation, so sessions can be created and destroyed without significant cost. Always xref:_close_sessions[close sessions] when you are done with them.
```
```suggestion
Session creation is a lightweight operation, so sessions can be created and destroyed without significant cost. 
Always xref:_close_sessions[close sessions] when you are done with them.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+[#managed-transactions]
+== Run a managed transaction
+
+A transaction can contain any number of queries. As Neo4j is <<ACID>> compliant, *queries within a transaction will either be executed as a whole or not at all*: you cannot get a part of the transaction succeeding and another failing. Use transactions to group together related queries which work together to achieve a single _logical_ database operation.
```
```suggestion
A transaction can contain any number of queries. 
As Neo4j is <<ACID>> compliant, *queries within a transaction are either executed as a whole or not at all*: you cannot get a part of the transaction succeeding and another failing. 
Use transactions to group together related queries which work together to achieve a single _logical_ database operation.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+A transaction can contain any number of queries. As Neo4j is <<ACID>> compliant, *queries within a transaction will either be executed as a whole or not at all*: you cannot get a part of the transaction succeeding and another failing. Use transactions to group together related queries which work together to achieve a single _logical_ database operation.
+
+A managed transaction is created with the methods link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_read[`Session.execute_read()`] and link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_write[`Session.execute_write()`], depending on whether you want to retrieve data from the database or alter it. Both methods take a _transaction f ...
```
```suggestion
A managed transaction is created with the methods link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_read[`Session.execute_read()`] and link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.execute_write[`Session.execute_write()`], depending on whether you want to retrieve data from the database or alter it. 
Both methods take a _transaction function_ callback *(1)* and an arbitrary number of positional and keyword arguments *(2)* which are handed down to the transaction function. 
The transaction function *(3)* is responsible for actually carrying out the queries and processing the result. 
Queries are specified with the link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Transaction.run[`Transaction.run()`] method *(4)*, which returns a link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Result[`Result`] object. 
You can then xref:result#manipulate-result[process the result] *(5)* using any of the `Result` methods, or simply casting it to list.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+        print(person.data())  # obtain dict representation
+----
+
+*Do not hardcode or concatenate parameters directly into the query*. Use xref:query-simple#query-parameters[query parameters] instead, both for performance and security reasons.
```
```suggestion
*Do not hardcode or concatenate parameters directly into the query*. 
Use xref:query-simple#query-parameters[query parameters] instead, both for performance and security reasons.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+*Do not hardcode or concatenate parameters directly into the query*. Use xref:query-simple#query-parameters[query parameters] instead, both for performance and security reasons.
+
+*Transaction functions should never return the `Result` object directly*. Instead, always xref:result.adoc[process the result] in some way; at minimum, cast it to list. Within a transaction function, a `return` statement results in the transaction being committed, while the transaction is automatically rolled back if an exception is raised.
```
```suggestion
*Transaction functions should never return the `Result` object directly*. 
Instead, always xref:result.adoc[process the result] in some way; at minimum, cast it to list. 
Within a transaction function, a `return` statement results in the transaction being committed, while the transaction is automatically rolled back if an exception is raised.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+    main()
+----
+
+Should a transaction fail for a reason that the driver deems transient, it will automatically retry to run the transaction function (with an exponentially increasing delay). For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know upfront how many times they are going to be executed. In practice, this ...
```
```suggestion
Should a transaction fail for a reason that the driver deems transient, it automatically retries to run the transaction function (with an exponentially increasing delay). 
For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know upfront how many times they are going to be executed. 
In practice, this means that you should not edit nor rely on globals, for example. 
Note that although transactions functions might be executed multiple times, the queries inside it are always run only once.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+The decorator link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.unit_of_work[`unit_of_work()`] allows to exert further control on transaction functions. It allows to specify:
+
+- a transaction timeout (in seconds). Transactions that run longer will be terminated by the server. The default value is set on the server side.
```
```suggestion
- a transaction timeout (in seconds). 
Transactions that run longer are terminated by the server. 
The default value is set on the server side.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+The decorator link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.unit_of_work[`unit_of_work()`] allows to exert further control on transaction functions. It allows to specify:
+
+- a transaction timeout (in seconds). Transactions that run longer will be terminated by the server. The default value is set on the server side.
+- a dictionary of metadata that gets attached to the transaction. These metadata get logged in the server `query.log`, and are visible in the output of the `SHOW TRANSACTIONS` Cypher command. Use this to tag transactions.
```
```suggestion
- a dictionary of metadata that gets attached to the transaction. 
These metadata get logged in the server `query.log`, and are visible in the output of the `SHOW TRANSACTIONS` Cypher command. 
Use this to tag transactions.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+[#explicit-transactions]
+== Run an explicit transaction
+
+You can achieve full control over transactions by manually beginning one with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.begin_transaction[`Session.begin_transaction()`]. You run queries inside an explicit transaction with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Transaction.run[`Transaction.run()`].
```
```suggestion
You can achieve full control over transactions by manually beginning one with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Session.begin_transaction[`Session.begin_transaction()`]. 
You run queries inside an explicit transaction with the method link:{neo4j-docs-base-uri}/api/python-driver/current/api.html#neo4j.Transaction.run[`Transaction.run()`].
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+=== Database selection
+
+It is recommended to *always specify the database explicitly* with the `database` parameter, even on single-database instances. This allows the driver to work more efficiently, as it does not have to resolve the home database first. If no database is given, the link:{neo4j-docs-base-uri}/operations-manual/current/manage-databases/introduction#manage-databases-default[default database] set in the  ...
```
```suggestion
It is recommended to *always specify the database explicitly* with the `database` parameter, even on single-database instances. 
This allows the driver to work more efficiently, as it does not have to resolve the home database first. 
If no database is given, the link:{neo4j-docs-base-uri}/operations-manual/current/manage-databases/introduction#manage-databases-default[default database] set in the Neo4j instance settings is used.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+=== Request routing
+
+In a cluster environment, all sessions are opened in write mode, routing them to the leader. You can change this by explicitly setting the `default_access_mode` parameter to either `neo4j.READ_ACCESS` or `neo4j.WRITE_ACCESS`. Note that `.execute_read()` and `.execute_write()` automatically override the session's default access mode.
```
```suggestion
In a cluster environment, all sessions are opened in write mode, routing them to the leader. 
You can change this by explicitly setting the `default_access_mode` parameter to either `neo4j.READ_ACCESS` or `neo4j.WRITE_ACCESS`. Note that `.execute_read()` and `.execute_write()` automatically override the session's default access mode.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+[#access-control-note]
+[NOTE]
+====
+Although executing a _write_ query under read mode will likely result in a runtime error, *you should not rely on this for access control.* The difference between the two modes is that _read_ transactions will be routed to any node of a cluster, whereas _write_ ones will be directed to the leader. Still, depending on the server version and settings, the server might allow none, some, or all _writ ...
```
```suggestion
Although executing a _write_ query in read mode likely results in a runtime error, *you should not rely on this for access control.* 
The difference between the two modes is that _read_ transactions are routed to any node of a cluster, whereas _write_ ones are directed to the leader. 
Still, depending on the server version and settings, the server might allow none, some, or all _write_ statements to be executed even in _read_ transactions.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+[#impersonation]
+=== Run queries as a different user (impersonation)
+
+You can execute a query under the security context of a different user with the parameter `impersonated_user`, specifying the name of the user to impersonate. For this to work, the user under which the `Driver` was created needs to have the link:{neo4j-docs-base-uri}/cypher-manual/current/access-control/dbms-administration#access-control-dbms-administration-impersonation[appropriate permissions]. ...
```
```suggestion
You can execute a query under the security context of a different user with the parameter `impersonated_user`, specifying the name of the user to impersonate. 
For this to work, the user under which the `Driver` was created needs to have the link:{neo4j-docs-base-uri}/cypher-manual/current/access-control/dbms-administration#access-control-dbms-administration-impersonation[appropriate permissions]. 
Impersonating a user is cheaper than creating a new `Driver` object.
```

AlexicaWright commented on python-manual/modules/ROOT/pages/transactions.adoc:
```diff
@@ -0,0 +1,306 @@
+
+== Close sessions
+
+Each connection pool has *a finite number of sessions*, so if you open sessions without ever closing them, your application could run out of them. It is thus recommended to create sessions using the `with` statement, which automatically closes them when the application is done with them. When a session is closed, it is returned to the connection pool to be later reused.
```
```suggestion
Each connection pool has *a finite number of sessions*, so if you open sessions without ever closing them, your application could run out of them. 
It is thus recommended to create sessions using the `with` statement, which automatically closes them when the application is done with them. 
When a session is closed, it is returned to the connection pool to be later reused.
```

AlexicaWright reviewed (commented):
Your magnum opus is brilliant, @stefano-ottolenghi ! But to be consistent with how we work in other manuals, you need to implement one-sentence-per-line. Makes it so much easier to work with. Also, always use present tense.
One last note would be that this is a rather large PR..

robsdedude reviewed (approved):
:nerd_face: :book:

stefano-ottolenghi committed:
one sentence per line + minor changes from jessica

stefano-ottolenghi committed:
some glitches
